File: Ballast.Core/src/Client/models/board-type.ts

export interface IBoardType { 
    value: number;
    name: string;
    centerOrigin: boolean;
}

export class BoardType implements IBoardType {

    public static Rectangle: BoardType = new BoardType({ value: 0, name: 'Rectangle', centerOrigin: false });
    public static RegularPolygon: BoardType = new BoardType({ value: 1, name: 'RegularPolygon', centerOrigin: true });
    
    public readonly value: number;
    public readonly name: string;
    public readonly centerOrigin: boolean;

    private constructor(state: IBoardType) {
        this.value = state.value;
        this.name = state.name;
        this.centerOrigin = state.centerOrigin;
    }

    public static list(): BoardType[] {
        return [
            BoardType.Rectangle,
            BoardType.RegularPolygon
        ];
    }

    public equals(BoardType: BoardType) {
        if (!BoardType) {
            return false;
        }
        return this.value == BoardType.value;
    }

    public static fromObject(object: IBoardType): BoardType {           
        let item = !!object && BoardType.fromValue(object.value);
        if (!item) {
            throw new Error(`Could derive board type from object (${object})`);
        }
        return item;
    }

    public static fromValue(value: number): BoardType {           
        let item = (value > -1) && BoardType.list().find(x => x.value == value);
        if (!item) {
            throw new Error(`Could derive board type from value (${value})`);
        }
        return item;
    }

    public static fromString(text: string): BoardType {           
        let item = !!text && BoardType.list().find(x => x.name.toLocaleLowerCase() == text.toLocaleLowerCase());
        if (!item) {
            throw new Error(`Could derive board type from text (${text})`);
        }
        return item;
    }

}

File: Ballast.Core/src/Server/Models/BoardType.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ballast.Core.Models
{

    public class BoardTypeState : StaticListTypeStateBase
    {
        public bool CenterOrigin { get; set; }
    }

    public class BoardType : StaticListTypeBase<BoardType>
    {

        public readonly static BoardType Rectangle = new BoardType(value: 0, name: nameof(Rectangle), centerOrigin: false);
        public readonly static BoardType RegularPolygon = new BoardType(value: 1, name: nameof(RegularPolygon), centerOrigin: true);

        public bool CenterOrigin { get; private set; }

        private BoardType(int value, string name, bool centerOrigin) : base(value, name)
        {
            CenterOrigin = centerOrigin;
        }

        public static IEnumerable<BoardType> List() => new[] {
            BoardType.Rectangle,
            BoardType.RegularPolygon
        };

        public static BoardType FromValue(int value) =>
            BoardType.List().Single(x => x.Value == value);

        public static BoardType FromString(string name) =>
            BoardType.List().Single(x => x.Name.ToLowerInvariant() == name.ToLowerInvariant());
            
        public static implicit operator BoardType(BoardTypeState state) =>
            BoardType.FromValue(state.Value);

    }
    
}